### PR TITLE
Shutdown: fixes to solve issues in the shutdown sequence

### PR DESCRIPTION
**Pull request by HaseenaSainul**

### CODE REVIEW

File: Source/WPEFramework/PluginServer.cpp [hunk 1]
```diff
@@ -701,14 +701,16 @@ namespace PluginHost
     {
     }
 
-	void Server::Notification(const ForwardMessage& data)
+    void Server::Notification(const ForwardMessage& data)
     {
-        _controller->ClassType<Plugin::Controller>()->Notification(data);
-		#ifdef RESTFULL_API
+        if (_controller->ClassType<Plugin::Controller>() != nullptr) {
```
Comment (pwielders) [inline Source/WPEFramework/PluginServer.cpp]: This is a bit scary. It probably indicates that there is a plugin that is not Deactivated or is not adhering to the deactivate. I would suggest placing an ASSERT here and check the callstack if this happens to correct the plugin that is issueing this notification.
If a PLugin is not properly shut down and is still firing notifications it might even be that _controller is invalid and even the call _controller->ClassType<> is failing.

File: Source/WPEFramework/PluginServer.cpp [hunk 2]
```diff
@@ -701,14 +701,16 @@ namespace PluginHost
...
+            _controller->ClassType<Plugin::Controller>()->Notification(data);
+        }
+#ifdef RESTFULL_API
         string result;
         data.ToString(result);
         _controller->Notification(result);
```
Comment (pwielders) [inline Source/WPEFramework/PluginServer.cpp]: If the above situation occurs (no Controller available) this should probably also be protected.

File: Source/WPEFramework/PluginServer.cpp (hunk 1 above)
Comment (HaseenaSainul) [reply Source/WPEFramework/PluginServer.cpp]: Yea this is happening currently we are deactivating the plugin in the Alphabetic order Z to A. So Controller plugin is get deactivated before deactivating Compositor Plugin.

File: Source/WPEFramework/PluginServer.cpp (hunk 2 above)
Comment (HaseenaSainul) [reply Source/WPEFramework/PluginServer.cpp]: Sure will ensure this as well

File: Source/WPEFramework/PluginHost.cpp [hunk 3]
```diff
@@ -53,6 +57,35 @@ namespace PluginHost {
...
+
+    private:
+        virtual uint32_t Worker() override
+        {
+            CloseDown();
+            Block();
+            return (Core::infinite);
+        }
+    };
+    ExitHandler *exitHandler = nullptr;
```
Comment (pwielders) [inline Source/WPEFramework/PluginHost.cpp]: Move all CloseDown logic in this class (including the CloseDown code). Make it the responsibility of this class to "properly" shutdown. You can Than at the end of the main, in stead of checking another variable (exitHandler), just call "Destruct" (static method). This Destruct Method can on this class checks if the internal method exitHandler is set, a closeDown already in progress (and just wait for the Thread to reach completion) and then delete it or if the Variable is not set, just call CloseDown() and wait for Completion. 
Then you can also protect this variable (exitHandler), as currently you might cause a race condition. 
What happens if a signal arrives (QUIT/TERM) and you are already at 629, so CloseDown already occured....
Lets give this class a static method called Start() that fires the thread and is the only one doing the CloseDown.
Than a second static method called Completed() that WaitsForTheThread ToCompleet

File: Source/WPEFramework/PluginServer.cpp (hunk 1 above)
Comment (pwielders) [reply Source/WPEFramework/PluginServer.cpp]: Ohhh... Good analyses again!!! Since the Controller already has a Special status (it is a aggregated member of the server), let, by definition, not stop this plugin if we do the Close down of the server. Lets close it down after all other plugins have been stopped. This logic should go into the method where we actually "Stop"all plugins (Skip controller, by comparing it to the _controller member. and than if all plugins are deactivated deactivate this one. Than still it is a good idea to leave a DumpCallStack ere in case a plugin misbehaves .

File: Source/WPEFramework/PluginHost.cpp [hunk 4]
```diff
@@ -65,46 +140,18 @@ namespace PluginHost {
             fprintf(stderr, "Signal received %d.\n", signo);
         }
 
-        if (signo == SIGTERM) {
-            if (Core::WorkerPool::IsAvailable() == true) {
-				Core::WorkerPool::Instance().Stop();
-			}
+        if ((signo == SIGTERM) || (signo == SIGQUIT)) {
+            WPEFramework::PluginHost::ExitHandler* exitHandler = new WPEFramework::PluginHost::ExitHandler();
```
Comment (pwielders) [inline Source/WPEFramework/PluginHost.cpp]: Can you move these lines 144/145 into a static method in the ExitHandler class as well (make the constructors of this ExitHandler class private so yu can not new it and only use this static Construct.
The reason for it being that you need to protect this creation. If the internal static is already "set"you do not need to start another one (so nothing to be done, a previous signal already created this). Also if the other static public method, the Destructwas already called, you also do not need to create a new exitHandler.
So we need to check for some additional conditions there.
These issues would occure if before, during or after the closing down you issue the SIGTEM/SIGQUIT signal multiple times.

Comment (HaseenaSainul) [reply Source/WPEFramework/PluginHost.cpp]: Sure, will arrange

File: Source/WPEFramework/PluginServer.cpp [hunk 5]
```diff
@@ -701,14 +708,19 @@ namespace PluginHost
     {
     }
 
-	void Server::Notification(const ForwardMessage& data)
+    void Server::Notification(const ForwardMessage& data)
     {
+        if ((_controller.IsValid() == false) || (_controller->ClassType<Plugin::Controller>() == nullptr)) {
+            DumpCallStack();
+        }
+
```
Comment (pwielders) [inline Source/WPEFramework/PluginServer.cpp]: Shouldent we do an else { here ? If this situation occures, which is a "real issue", you gnerate a callstack but would not call this as this would crash the system due to nullptr access

File: Source/plugins/VirtualInput.h [hunk 6]
```diff
@@ -175,7 +175,6 @@ namespace PluginHost {
                 }
 
                 ChangeIterator removed(removedKeys);
-                _parent.MapChanges(removed);
```
Comment (pwielders) [inline Source/plugins/VirtualInput.h]: I would like to investigate this one a bit further. Cause now I think we never get removal of keys mapping notifications. Why did you remove this one ?

Comment (HaseenaSainul) [reply Source/plugins/VirtualInput.h]: We are clearing the keymap table from the destructor of VirtualInput from here : https://github.com/WebPlatformForEmbedded/Thunder/blob/master/Source/plugins/VirtualInput.cpp#L193
This will invoke the destructor of KeyMap: https://github.com/WebPlatformForEmbedded/Thunder/blob/master/Source/plugins/VirtualInput.h#L164. VirtualInput is its parent: https://github.com/WebPlatformForEmbedded/Thunder/blob/master/Source/plugins/VirtualInput.h#L159 and it is already in destruction state. Hence the line cause segfaults : https://github.com/WebPlatformForEmbedded/Thunder/blob/master/Source/plugins/VirtualInput.h#L178

Comment (pwielders) [reply Source/plugins/VirtualInput.h]: As usual a very good analyses again!! Now lets think about how to solve this as now I think we are also removing notification of map changes if triggered through the PLugin (so not during destruction)

Comment (HaseenaSainul) [reply Source/plugins/VirtualInput.h]: Sure, I will try to manage it.

**Comment by goruklu** [comment]
@pwielders @HaseenaSainul grea! looks like these commits will fix https://github.com/WebPlatformForEmbedded/Thunder/issues/207

Do you know if they will help with https://github.com/WebPlatformForEmbedded/Thunder/issues/218 as well ? Currently if OCDM WPEProcess crashes, that requires restarting the client (WPE Webkit) process...

**Comment by HaseenaSainul** [comment]
@goruklu, yes, it will address https://github.com/WebPlatformForEmbedded/Thunder/issues/207.
I have to check https://github.com/WebPlatformForEmbedded/Thunder/issues/218, separately.

File: Source/plugins/VirtualInput.h [hunk 7]
```diff
@@ -333,6 +338,12 @@ namespace PluginHost {
         virtual ~VirtualInput();
 
     public:
```
Comment (pwielders) [inline Source/plugins/VirtualInput.h]: \suggest to make this a protected method. Only derived classes now what to do with this. Also think that you can move the ClearRemovedKeys into this methods or make the ClearRemovedKeys private. Users of this class should not exercise these methods.
